app/channels.go

```diff
@@ -45,6 +47,18 @@ type Channels struct {
 	Compliance       einterfaces.ComplianceInterface
 	DataRetention    einterfaces.DataRetentionInterface
 	MessageExport    einterfaces.MessageExportInterface
+
+	// These are used to prevent concurrent upload requests
+	// for a given upload session which could cause inconsistencies
+	// and data corruption.
+	uploadLockMapMut sync.Mutex
+	uploadLockMap    map[string]bool
+
+	imgDecoder *imaging.Decoder
+	imgEncoder *imaging.Encoder
```

Comment:
I could see these go beyond channels at some point given they are generic enough but I guess it's fine to leave them closer to where they are currently used.

Reply:
Yep, agree. I am starting small, and we can expand to more common services as and when products need it.